Between the releases of Deadwood 2.9.05 and 2.9.06, there was an issue with
bogus "empty" packets slowing down Deadwood when trying to process AAAA
queries (see the "mt_packet" patch in update/2.9.06 for the full details).

Unfortunately, while some DNS servers incorrectly return a REFUSED answer
not because they do not like AAAA queries, but because they are incorrectly
configured (for example, right now one of archive.org's three DNS servers
gives us a REFUSED answer).

This patch minimizes the issue by *only* creating synthetic "not there"
answers when an upstream DNS server gives us a REFUSED answer and the
type of record asked for was an AAAA record.

--- deadwood-3.0.02/src/DwRecurse.c	2011-01-29 14:55:10.000000000 -0600
+++ deadwood-3.0.03/src/DwRecurse.c	2011-04-15 12:12:31.000000000 -0500
@@ -1,4 +1,4 @@
-/* Copyright (c) 2009-2010 Sam Trenholme
+/* Copyright (c) 2009-2011 Sam Trenholme
  *
  * TERMS
  *
@@ -1559,8 +1559,15 @@
         dns_details *view = 0;
         dw_str *out = 0;
         int type = 0;
+	int_fast32_t qtype = 0;
 
-        if(in->len == 7) { /* Blank packet */
+	qtype = dw_fetch_u16(query,-1);
+
+	if(qtype == -1) {
+		return 0;
+	}
+
+        if(in->len == 7 && qtype == RR_AAAA) { /* Blank packet */
                 return dwx_synth_notthere(query);
         }
 
